director/projects/api/views/snapshots.py

```diff
@@ -0,0 +1,258 @@
+    def retrieve(self, request: Request, pk: int, number: int) -> Response:
+        """
+        Retrieve a snapshot of the project.
+
+        Returns details of the snapshot, including date/time `created`
+        and `completed`, and any `tag`.
+        """
+        # Check that the user has VIEW permissions for the project
+        self.get_project(request.user, pk=pk)
+        if not self.has_permission(ProjectPermissionType.VIEW):
+            raise PermissionDenied
+
+        snapshot = get_object_or_404(Snapshot, project=pk, version_number=number)
+        serializer = SnapshotSerializer(snapshot)
+        return Response(serializer.data)
+
+    @action(detail=True, methods=["get"], url_path="(?P<path>.+)")
+    def retrieve_file(
+        self, request: Request, pk: int, number: int, path: str
+    ) -> Response:
+        """
+        Retrieve a file within the snapshot of the project.
+
+        Returns the content of the file, either as is, or converted
+        to the desired format as specified in the `format` query parameter
+        or the `Accept` header.
+        """
+        # Check that the user has VIEW permissions for the project
+        project = self.get_project(request.user, pk=pk)
+        if not self.has_permission(ProjectPermissionType.VIEW):
+            raise PermissionDenied
+
+        snapshot = get_object_or_404(Snapshot, project=pk, version_number=number)
+
+        file_path = snapshot_path(snapshot, path)
+        if not os.path.exists(file_path) or os.path.isdir(file_path):
+            raise NotFound
+
+        # Determine the desired format
+        format = request.query_params.get("format")
+        if format is not None:
+            format_id = ConversionFormatId.from_id(format)
+        else:
+            accept = request.META.get("HTTP_ACCEPT")
+            if accept is not None and accept != "*/*":
+                media_types = [token.strip() for token in accept.split(",")]
+                format_id = ConversionFormatId.from_mimetype(media_types[0])
+            else:
+                format_id = None
+
+        raw = request.query_params.get("raw")
+        if raw or format_id is None:
+            # Respond with the file
+            response_path = file_path
+        else:
+            # Generate file for the format if necessary and return it
+            response_path = snapshot_path(
+                snapshot, ".cache/{}.{}".format(path, format_id.value.default_extension)
+            )
+            if not os.path.exists(response_path):
+                ConverterFacade(settings.STENCILA_BINARY).convert(
+                    input_data=ConverterIo(ConverterIoType.PATH, file_path),
+                    output_data=ConverterIo(ConverterIoType.PATH, response_path),
+                    context=ConverterContext(standalone=True),
+                )
+        file = open(response_path, "rb")
+
+        # Return "binary" files as attachments
+        if format_id is None or format_id.value.is_binary:
```

Comment:
I'm not 100% sure on this but I think you can use a `FileResponse` in both cases.
```python
as_attachment = format_id is None or format_id.value.is_binary
response = FileResponse(file, as_attachment=as_attachment)

if hosts:
   …
```

I'm not sure from the docs but I think you should be able to pass in `content_type=format_id.value.mimetypes[0]` to the constructor too. If not, `response["Content-Type"] = …` should do it.

Reply:
Because a file could be any type of file, not just those that `ConversionFormatId` knows about, `FileResponse` has the advantage that it will determine that.